### PR TITLE
Add chat localization support

### PR DESCRIPTION
**Pull request by maxnorth**
This PR only includes the infrastructure for localization, but does not yet contain actual Burmese translations. Will add these in a future PR once they have been provided.

### CODE REVIEW

**Review by blueelvis — DISMISSED** [review]
Please review the changes.

File: api/src/chat/dialogs/myanmar/disaster-assessment/review/reviewAffectedPeopleDialog.ts [hunk 1]
```diff
@@ -26,55 +24,60 @@ export const reviewAffectedPeopleDialog: RcdaChatDialog = rcdaChatDialog(
         references: () => [askAffectedPeople.askAffectedPeopleDialog]
     });
 
-function createAffectedPeopleReviewCard(session: RcdaTypedSession):Message {
+const editAction = "edit";
```
Comment (blueelvis) [inline api/src/chat/dialogs/myanmar/disaster-assessment/review/reviewAffectedPeopleDialog.ts]: Let's ensure that the declarations are always at the top after imports. Easier to find :)

File: api/src/chat/utils/rcdaChatDialog.ts [hunk 2]
```diff
@@ -13,19 +16,19 @@ export default function rcdaChatDialog<TDialogData, TDependencies>(
 export function rcdaChatDialogStateful<TDialogData, TDependencies>(
```
Comment (blueelvis) [inline api/src/chat/utils/rcdaChatDialog.ts]: Just curious but what is the purpose of this function Max?

**Review by blueelvis — CHANGES_REQUESTED** [review]

File: api/src/chat/localization/RcdaChatLocalizer.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import { RcdaCommonTextEnglish, RcdaMyanmarTextEnglish } from "@/chat/localization/RcdaTextEnglish";
+import { RcdaCommonTextBurmese, RcdaMyanmarTextBurmese } from "@/chat/localization/RcdaTextBurmese";
+import { RcdaTypedSession } from "@/chat/utils/rcda-chat-types";
+
+export type RcdaCommonText = RcdaCommonTextEnglish;
+export type RcdaMyanmarText = RcdaMyanmarTextEnglish;
```
[the remark this answers is not in the document]
Comment (maxnorth) [reply api/src/chat/localization/RcdaChatLocalizer.ts]: Not sure about this, the English implementation serves as the base type, RcdaMyanmarTextBurmese (and any other potential future languages) inherits from that.

File: api/src/chat/localization/RcdaChatLocalizer.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    "en": RcdaMyanmarTextEnglish,
+    "my": RcdaMyanmarTextBurmese
+};
+
+function getLocalizedText<TResult>(
+    textLocaleMap: {[key:string]: new() => TResult}, 
+    name: string,
+    session: RcdaTypedSession): TResult {
+    
+    let locale = session.preferredLocale().split("-")[0].toLowerCase();
```
[the remark this answers is not in the document]
Comment (maxnorth) [reply api/src/chat/localization/RcdaChatLocalizer.ts]: Used this approach to be more versatile regarding locales represented as 'en' vs 'en-US' which is also common. Regardless of the format, split("-")[0] will correctly return 'en', so this is effectively a way to normalize it.

Just curious, where did you find this sample? Tried this approach and it doesn't work for me. I do something very similar in api/src/chat/dialogs/selectLanguageDialog.ts, though.

File: api/src/chat/dialogs/myanmar/disaster-assessment/questions/askUserInfo.ts [hunk 5]
```diff
@@ -1,27 +1,29 @@
...
-            Prompts.choice(session, "What is the setting?", GeographicalSettings.map(x => x.name.en), { listStyle: ListStyle.button });
+            Prompts.choice(session, localizer.mm.askGeographicalSettingType, Object.values(localizer.mm.geographicalSettings), { listStyle: ListStyle.button });
         },
-        ({ session, result }) => {
-            session.conversationData.mm.geographicalSettingId = GeographicalSettings.find(x => x.name.en === result.response.entity).id;
+        ({ session, localizer, result }) => {
+            console.log("!!!!!!!!!"+JSON.stringify(result));
+            session.conversationData.mm.geographicalSettingId = Object.keys(localizer.mm.geographicalSettings)[result.index] as MyanmarGeographicalSettings;
+            console.log(session.conversationData.mm.geographicalSettingId);
+            //TODO this isn't working
```
[the remark this answers is not in the document]
Comment (maxnorth) [reply api/src/chat/dialogs/myanmar/disaster-assessment/questions/askUserInfo.ts]: A value wasn't getting assigned to conversationData.mm.geographicalSettingsId. Found the issue, I was using 'result.index', but I needed to use 'result.response.index'. Have a fix coming for this.

File: api/src/chat/dialogs/myanmar/disaster-assessment/selectFormSectionDialog.ts [hunk 6]
```diff
@@ -9,49 +9,49 @@ export const selectFormSectionDialog = rcdaChatDialog(
...
+                retryPrompt: localizer.mm.invalidChoicePromptRetry
             });
         },
-        ({ session, result }: RcdaChatStep<IPromptChoiceResult>) => {
+        ({ session, localizer, result }: RcdaChatStep<IPromptChoiceResult>) => {
             const selection =  result.response.entity;
-            if (selection === "People") {
+            if (selection === localizer.mm.reportSectionNamePeople) {
                 session.beginDialog(askAffectedPeopleDialog.id);
                 return;
```
[the remark this answers is not in the document]
Comment (maxnorth) [reply api/src/chat/dialogs/myanmar/disaster-assessment/selectFormSectionDialog.ts]: The return is just to end the execution immediately. Once we've called beginDialog, there's no work left to do, so no sense letting it check the other if statements. Technically we could remove this without changing the behavior, but I like that this is clear that the other if statements are ignored.

File: api/src/chat/dialogs/myanmar/disaster-assessment/selectFormSectionDialog.ts [hunk 7]
```diff
@@ -9,49 +9,49 @@ export const selectFormSectionDialog = rcdaChatDialog(
...
-                retryPrompt: "Sorry, I didn't understand that. Please select one of the listed options."
+                retryPrompt: localizer.mm.invalidChoicePromptRetry
             });
         },
-        ({ session, result, next }) => {
+        ({ session, localizer, result, next }) => {
             const selection = (<IPromptChoiceResult>result).response.entity;
-            if (selection === "Yes") {
+            if (selection === localizer.common.yes) {
                 session.beginDialog(selectFormSectionDialog.id);
```
[the remark this answers is not in the document]
Comment (maxnorth) [reply api/src/chat/dialogs/myanmar/disaster-assessment/selectFormSectionDialog.ts]: Good catch, will update. Can also replace next() with endDialog(). Probably has the same effect, but I think it's semantically more appropriate to indicate that the intent is to end.

File: api/src/chat/dialogs/myanmar/disaster-assessment/selectFormSectionDialog.ts [hunk 8]
```diff
@@ -9,49 +9,49 @@ export const selectFormSectionDialog = rcdaChatDialog(
...
+                localizer.common.no
             ], {
                 listStyle: ListStyle.button,
-                retryPrompt: "Sorry, I didn't understand that. Please select one of the listed options."
+                retryPrompt: localizer.mm.invalidChoicePromptRetry
             });
         },
-        ({ session, result, next }) => {
+        ({ session, localizer, result, next }) => {
             const selection = (<IPromptChoiceResult>result).response.entity;
```
[the remark this answers is not in the document]
Comment (maxnorth) [reply api/src/chat/dialogs/myanmar/disaster-assessment/selectFormSectionDialog.ts]: Entity is just the property containing the response text. Open to other approaches, if you have a preferred alternative.

File: api/src/chat/utils/rcdaChatDialog.ts (hunk 2 above)
Comment (maxnorth) [reply api/src/chat/utils/rcdaChatDialog.ts]: This allows for a dialog to define a dialogData constructor. The constructor will be run once per dialog, before the first waterfall dialog step. session.dialogData for that dialog will then be strongly typed according to the result of the dialogData constructor. I may later collapse this into the normal rcdaChatDialog method as an overload.

**Comment by maxnorth** [comment]
@blueelvis Responded to all comments and pushed changes.

**Comment by maxnorth** [comment]
@blueelvis, I'm going to land this so we can move forward on other PR's, but please review my responses above and let me know if you have any additional change requests, I'll address them in a separate PR.

**Pull request merged**